### ISSUE TITLE
new method in Recur to get dates as a java.util.stream.Stream

### ISSUE THREAD

**Issue by fmossott** [ical4j/ical4j#612]
**Is your feature request related to a problem? Please describe.**
Recur.getDates has better performances that iterating using Recur.getNextDate, however the drawback is that it produces all dates in memory.

**Describe the solution you'd like**
A new method (e.g. Recur.getDatesStream) that returns a java.util.stream.Stream<T>
The method could be similar to getDates, but it won't actually calculate all the dates that would be calculated as they are consumed by the stream.
To build the Stream it could use java.util.stream.Stream#iterate

**Describe alternatives you've considered**
Using getDates().stream() still creates the list of dates before streaming, keeping them all in memory
Tried to create the Steam using java.util.stream.Stream#iterate and calling getNextDate, but it has bad performances and fails in some scenarios

**Comment by benfortuna** [comment]
It is an interesting idea. 

I think the challenge may be that the way we calculate recurrences requires applying the rules in sequence to all of the dates in the given range. 

So most of the processing cannot be done lazily, or in parallel, which seems to be where you get the most benefit from streams.

It may be possible to consider processing in batches for larger date ranges, but will need to look at whether that would work for all recurrence rules.

**Comment by fmossott** [comment]
Streams elements are not produced when you create the Stream, but when they are consumed by the final operation.
Stream.iterate returns a sequential stream, however it can be converted to parallel just calling the parallel() method (the consumer can call it if needed).
This already provides a basic parallelism support, that can be eventually improved implementing the trySplit on the Spliterator.
Stream.iterate() is actually available since Java 9 and I see you support also Java 8, so it could be required to implement a Spliterator anyway.

Of course you need to keep the status in memory (I usually use an inner class that implement the next / hasNext methods and  keeps the status in its fields)

Looking at the current getDates, you basically have 2 nested loops, one on increment() for the interval, and one on elements returned by getCandidates().
Without changing too much the logic, it should be possible to keep the candidates in a list in the inner class, returning them one by one and then increment and reload the next candidates.

The part that breaks the approach is the final sort of the list that would require to have all results available, why is that needed?

**Comment by benfortuna** [comment]
Thanks, I just had a closer look at Spliterator based on your comments, and like you said I was able to copy/paste the code from getDates() pretty much unchanged and it worked!

I don't have trySplit() implemented yet (not sure how to do that yet), but I should be able to release basic functionality after some more tests and clean up of code (I think this may help improve the overall code base too).

**Comment by fmossott** [comment]
I've tried running with your latest code.
In some situations I'm getting ArrayOutOfBound when tryAccept calls action.accept(), it can call the accept only once on each invocation, that requires to park candidates and the index of last returned candidate in the fields.

you could also consider extending Spliterators.AbstractSpliterator<T> instead of implementing Spliterator<T>

**Comment by benfortuna** [comment]
Yeah I realised it was calling accept () multiple times.. I'll have another look soon to fix this.

**Comment by fmossott** [comment]
If you prefer to hold it, I've a working implementation, I've started the internal process to get approval by my company in order to share it with you in a pull request. Not sure how long it will take to get the approval though.

**Comment by benfortuna** [comment]
Ok, that would be great if you get approval. Otherwise let me know and I'll put some more effort into it.

Much appreciated.

**Comment by benfortuna** [comment]
Note, I've made some more changes that I think should fix the date stream issues.

**Comment by fmossott** [comment]
I've finally got the approval, but you code works just fine now.
If you are curious, the implementation I did is here: https://github.com/HCL-TECH-SOFTWARE/ical4j/tree/feature/stream

**Comment by benfortuna** [comment]
Thanks for sharing the code. If you find the existing implementation is less performant feel free to suggest optimizations based on how you have implemented.

I did notice one difference in the Spliterator constructor:

```java
super(Long.MAX_VALUE, Spliterator.ORDERED | Spliterator.IMMUTABLE);
```

I currently don't sort the stream, but noticed you are applying the CANDIDATE_SORTER. Let me know if you think sorting is required (I've moved the expectation to the caller to sort).
